tools.rkt

```diff
@@ -305,6 +311,131 @@
+       (explore-loc expanded-node expanded-context))]))
+(define (explore-undo exp-loc)
+  (match exp-loc
+    [(explore-loc tree (explore-parent-context i siblings ch ctx))
+     (explore-loc (explore-node i ch (cons tree siblings)) ctx)]
+    [(explore-loc t 'X-TOP) (explore-loc t 'X-TOP)]))
+
+;; pretty-print functions
+(define (pp/qvars qvars vs)
+  (define (qv-prefix qv) (string-append " " (symbol->string qv) " = "))
+  (define qv-prefixes (and qvars (map qv-prefix qvars)))
+  (if qv-prefixes
+      (for-each (lambda (prefix v) (pprint/margin "" prefix v)) qv-prefixes vs)
+      (for-each (lambda (v) (pprint/margin "" " " v)) vs)))
+(define (pprint-choice s qvars)
+  (match s
+    [(pause st g)
+      (pp/qvars qvars (walked-term initial-var st))
+      (define cxs (walked-term (goal->constraints st g) st))
+      (unless (null? cxs)
+        (displayln "Constraints:")
+        (for-each (lambda (v) (pprint/margin "" " * " v)) cxs))
+      (when (null? cxs)
+        (displayln "No constraints")
+        (newline))]))
+(define (pprint-result s qvars)
+  (displayln "Result:")
+  (pp/qvars qvars (walked-term initial-var s)))
+(define (pprint-choices choices qvars)
+  (define chs (dropf choices state?))
+  (define results (takef choices state?))
+  (when (and (= 0 (length chs)) (null? results))
+    (printf "No more choices available. Undo to continue.\n"))
+  (unless (null? chs)
+    (printf "Number of Choices: ~a\n" (length chs))
+    (for-each (lambda (i s)
+                (printf (string-append "\nChoice ~s:\n") (+ i 1))
+                (pprint-choice s qvars))
+              (range (length chs)) chs))
+  (unless (null? results)
+    (printf "Number of results: ~a\n" (length results))
+    (for-each (lambda (st)
+                (pprint-result st qvars)
+                (newline))
+              results)))
+
+;; policy-print, policy-read, policy-done?
+(define (pp/explore-tree exp-loc qvars)
+  (define tree (explore-loc-tree exp-loc))
+  #| (printf "Tree: ~s\n" tree) |#
+  #| (printf "Context: ~s\n" (explore-loc-context exp-loc)) |#
+  (pprint-choices (explore-node-choices tree) qvars))
+(define (explore-tree-input)
+  (printf "\n[u]ndo, or choice number> \n")
+  (read))
+(define (explore-tree-finished? exp-loc)
+  (let* ([tree (explore-loc-tree exp-loc)]
+         [choices (explore-node-choices tree)]
+         [finished-index (index-where
+                           choices
+                           (lambda (x) (state? x)))]
+        [valid-index (exact-nonnegative-integer? finished-index)])
+    valid-index))
+
+(define (drive/policy step qvars policy-print policy-read policy-done? init-state)
+  (let loop ([s init-state])
+    (policy-print s qvars)
+    (unless (policy-done? s)
+      (let* ([input (policy-read)]
+             [tree (explore-loc-tree s)])
+        (loop
+          (cond
+            [(and (integer? input) (<= 1 input) (<= input (length (explore-node-choices tree))))
+            (explore-choice s step (- input 1))]
```

Comment:
Thanks, this looks good now.  Later, if we expand the actions an agent may take, we may have to give more structure to the commands coming from the policy.  But looking for `integer?` vs. `u` is fine for now.

The indentation should have another space.